Changed shorten to return tagged tuple when the length is shorter than the amount

paolodm opened the pull request:
Implementation for https://github.com/mykewould/crutches/issues/94. 
- As @duijf mentioned, we will return a tagged tuple instead of nil (in the scenario that the user passes an `amount < len(list)`). 
- I also delegated the ok function to @knrz's Towel library.

paolodm commented:
@knrz I am a quite new to elixir. I tried to find a solution to the following problem to no avail:

```
     Doctest failed: got UndefinedFunctionError with message undefined function Towel.Result.ok/1 (module Towel.Result is not available)
```

Can you help me understand why this is not working?

CyrusNuevoDia commented:
You need to add `{:towel, "~> 0.2"}` to the `mix.exs` file, then run `mix deps.get` in your shell. This pulls in my Towel library as a dependency and lets you use its modules.

paolodm commented:
Isn't that what this is?

https://github.com/mykewould/crutches/pull/107/files#diff-6023be6004fce4718dad3dafb576d258R29

Even after `mix deps.get`, I still get the same errors.

CyrusNuevoDia commented:
Oh crap. I didn't see that. Let me clone and test it.

duijf commented:
@paolodm, @knrz Towel doesn't have nested module names; `Towel.Result` doesn't exist, the module is called `Result`, therefore Elixir complains that `Towel.Result` is not available.

Also, Towel's `Result.ok/1` doesn't seem to unwrap tagged tuples, but wrap values in an `{:ok, value}` tuple (exactly the opposite of what I meant :)).

Most of the discussion with regard to using the error handling in a pipeline is currently moot though, since Elixir v1.2 gives us the [`with` statement](http://elixir-lang.org/getting-started/mix-otp/docs-tests-and-with.html#with).

I think the best way to proceed is:
- Throw out the Towel dependency, since Elixir provides a first class construct for what we needed it for (I don't want to introduce non-idiomatic error handling code even in older Elixir versions; people can write the functions themselves or use Towel if they feel like they need them, but it should be a conscious choice).
- Also return `{:ok, shortened_list}` on success?

paolodm commented:
Sounds reasonable. Updated with changes and ready to review again!

paolodm commented on mix.exs:
```diff
@@ -3,7 +3,7 @@ defmodule Crutches.Mixfile do
   def project do
     [app: :crutches,
-     version: "2.0.0",
+     version: "3.0.0",
```
Since the return type for List.shorten changed, this seems appropriate.

CyrusNuevoDia replied:
I think we should keep it at 2.0.0, because 
1. Reduce churn
2. We haven't pushed 2.0.0 to hex, so it's not like the previous changes to `List.shorten` were released.

duijf commented:
@paolodm, I agree with @knrz in keeping the version at 2.0.0.

Other than that, this looks ready for merging! :) Could you please change the version back and squash your commits into a single one?

paolodm committed:
Changed List.shorten to return a tagged tuple
Instead of just returning the shortened list, List.shorten will return a tagged tuple, i.e.:

`{:ok, [1, 2, 3]}` instead of just `[1, 2, 3]`
`{:error, "Amount to shorten by is greater than the length of the list"}` instead of `nil`

Resolves #94.

paolodm commented:
Put back the 2.0.0 version and squashed to 1 commit. 

Contributing to this project is a great way to learn Elixir. Let me know of any other issues you may have. Thanks!

duijf commented:
Thanks a lot, and will do! The same goes for you; if you find yourself writing a general function that others would benefit from, feel free to open an issue or send a PR!
